extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/ExcludePersistenceXmlConfigTest.java

```diff
@@ -0,0 +1,32 @@
+package io.quarkus.hibernate.orm;
+
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.hibernate.orm.enhancer.Address;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ExcludePersistenceXmlConfigTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClass(Address.class)
+                    .addAsManifestResource("META-INF/some-persistence.xml", "persistence.xml")
+                    .addAsResource("application-exclude-persistence-xml.properties", "application.properties"));
+
+    @Inject
+    EntityManager entityManager;
+
+    @Test
+    public void testPersistenceAndConfigTest() {
+        Assertions.assertNotNull(entityManager);
```

Comment:
I'm confused, shouldn't the test check that the matching entityManager was _not_ created?

What is the goal of the feature exactly?

Reply:
The goal is to filter out legacy persistence units coming from persistence.xml files. My use case is, that I depend on a jar file,  that contains a persistence.xml with options, that are incompatible with quarkus. I don't need this persistence unit, but need the classes from the jar. With this change, I can filter out the unwanted unit, but still use quarkus provided persistence units. That is the entity manager, that is asserted in the test

So maybe I should change the test, so that it asserts that the correct persistence unit from the application.properties is loaded.

The other idea to solve this use case for me, would be, to have an option, that completely disables persistence.xml parsing. But I thought this would be more flexible. Maybe someone has the use case of needing some units, but not others.

The only option that would be left to me if nothing like this is supported by quarkus, would be to unpack the dependency jar, remove the persistence.xml and repack it in the build process of the quarkus application. I would really like to not do that. :smile:

Reply:
> So maybe I should change the test, so that it asserts that the correct persistence unit from the application.properties is loaded.

Right, what I meant is that the test isn't t testing what you actually want.

I understand the general use case. My question about "exactly" is about what will you expect _exactly_ to happen?
The excluded persistence unit to not be booted at all?

What about the mapped entities in there, should they be ignored if they are not explicitly used by another persistence unit?
Should they be excluded from native-image reflection rules?
Should we not use bytecode enhancement on them?

Reply:
Before I answer, is the following understanding of the current behavior correct?

1. If no persistence.xml is found:
   - hibernate is configured inside the `application.properties`
   - all `@Entity` annotated classes from the current jar are added to this persistence unit
   - all `@Entity` annotated classes from other jars are added to this persistence unit, if that jar has a `META-INF/beans.xml` file
2. If one or more persistence.xml is found:
   - hibernate can not be configured anymore inside `application.properties`
   - all `@Entity` annotated classes from the current and referenced jars are added to the persistence units, that reference them
   - all `@Entity` annotated classes from the current and referenced jars, not referenced in any persistence unit, are not enhanced and treated as normal classes

If that is the case, then this change should behave the following:

3. If the exclude removes all persistence units from all persistence.xml files, we are in case 1. again
4. If some persistence units survive the exclusion process, we are in case 2. again

If my understanding of the current behavior is wrong, please give me some hints on how it works. :smile:

Reply:
I think that's correct - and sorry for digging, I need the details as I'm thinking at the implications this might have when we'll actually support multiple persistence units.

In particular the rules relating to `beans.xml` - you're right that's how it works today but I think it will need to change.

Could you:

1. improve the test to verify it's booting exactly the PU you expect (and no more)
2. improve the documentation on the new configuration properties to make it clear that these are experimental

Regarding marking this "experimental": I hope we'll actually be able to remove any reference and support for persistence.xml files at some time

Reply:
Will update the test. Should I change the property names to something like `quarkus.hibernate-orm.experimental.persistence-xml.exclude-persistence-unit` to make it clear that those are subject to change?

Reply:
Thanks! No need to use "experimental" in the name .. I'd just mention it on the description so that we can integrate it soon, and then possibly promote it to no-longer-experimental .

Reply:
Done. The CI failure seems to be unrelated.